api/api_ethereum.go

```diff
@@ -1455,11 +1455,19 @@ func EthDoEstimateGas(ctx context.Context, b Backend, args EthTransactionArgs, b
 		feeCap = args.MaxFeePerGas.ToInt()
 	}
 
-	state, _, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
 	if err != nil {
 		return 0, err
 	}
 	balance := state.GetBalance(*args.From) // from can't be nil
+	msg, err := deriveCallArgToMsg(b, &args, header, gasCap)
+	if err != nil {
+		return 0, err
+	}
+	requiredMinBalance := new(big.Int).Mul(new(big.Int).SetUint64(gasLimit), msg.EffectiveGasPrice(header, b.ChainConfig()))
+	if balance.Cmp(requiredMinBalance) < 0 {
+		balance = new(big.Int).Add(balance, requiredMinBalance)
+	}
 
 	executable := func(gas uint64) (bool, *blockchain.ExecutionResult, error) {
```

Comment:
As I read the code, I don't understand why estimateGas had been failing.
In EthDoCall and DoCall there already are lines like `AddBalance(sender, gasLimit * gasPrice)`. Please investigate why estimateGas fails in more depth.

Reply:
The call path is as follows:
`EstimateGas` -> `api/DoEstimateGas` -> `evm/DoEstimateGas`.
In `evm/DoEstimateGas`, the `test` function calls `DoCall`. Before doing this binary search iteration, there's a preceding validation here https://github.com/kaiachain/kaia/blob/dev/blockchain/evm.go#L176

Reply:
So users can pass in the `feePayer` field to the argument of `kaia_estimateGas` RPC? No, most SDKs won't do that. The estimateGas must work with the bare minimum of `{from, to, data, value}` regardless of other parameters.

By the way, the original example you posted is not related to the feePayer field nor fee delegated tx types.

Reply:
I think we should fix the line [dev/blockchain/evm.go#L176](https://github.com/kaiachain/kaia/blob/dev/blockchain/evm.go?rgh-link-date=2024-09-11T07%3A22%3A20Z#L176) you found.

Reply:
If the `feePayer` property is added, then this will be the member of `CallArgs`. Thus, no additional parameter position is placed. Just adding a new property in the object `{..}`. Empty is also allowed.

If this is added the line [dev/blockchain/evm.go#L176](https://github.com/kaiachain/kaia/blob/dev/blockchain/evm.go?rgh-link-date=2024-09-11T07%3A22%3A20Z#L176) seems still correct and we check the balance of `feePayer` instead.

Reply:
PTAL this diff (not pushed yet, but completed one)

```diff
diff --git a/api/api_ethereum.go b/api/api_ethereum.go
index 714fe948..bc40c4e5 100644
--- a/api/api_ethereum.go
+++ b/api/api_ethereum.go
@@ -1455,20 +1455,15 @@ func EthDoEstimateGas(ctx context.Context, b Backend, args EthTransactionArgs, b
 		feeCap = args.MaxFeePerGas.ToInt()
 	}
 
-	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	state, _, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
 	if err != nil {
 		return 0, err
 	}
 	balance := state.GetBalance(*args.From) // from can't be nil
-	msg, err := deriveCallArgToMsg(b, &args, header, gasCap)
-	if err != nil {
-		return 0, err
+	if !common.EmptyAddress(args.FeePayer) {
+		// overide the balance if the `feePayer` is given
+		balance = state.GetBalance(args.FeePayer)
 	}
-	requiredMinBalance := new(big.Int).Mul(new(big.Int).SetUint64(gasLimit), msg.EffectiveGasPrice(header, b.ChainConfig()))
-	if balance.Cmp(requiredMinBalance) < 0 {
-		balance = new(big.Int).Add(balance, requiredMinBalance)
-	}
-
 	executable := func(gas uint64) (bool, *blockchain.ExecutionResult, error) {
 		args.Gas = (*hexutil.Uint64)(&gas)
 		result, err := EthDoCall(ctx, b, args, blockNrOrHash, nil, b.RPCEVMTimeout(), gasCap)
diff --git a/api/api_public_blockchain.go b/api/api_public_blockchain.go
index 2d4b4095..884ce7c4 100644
--- a/api/api_public_blockchain.go
+++ b/api/api_public_blockchain.go
@@ -273,11 +273,6 @@ func (s *PublicBlockChainAPI) IsSenderTxHashIndexingEnabled() bool {
 	return s.b.IsSenderTxHashIndexingEnabled()
 }
 
-type CallMsg interface {
-	GetMsgIngredients() (*common.Address, types.AccessList, []byte)
-	ToMessage(globalGasCap uint64, baseFee *big.Int, intrinsicGas uint64) (*types.Transaction, error)
-}
-
 // CallArgs represents the arguments for a call.
 type CallArgs struct {
 	From                 common.Address  `json:"from"`
@@ -289,20 +284,13 @@ type CallArgs struct {
 	Value                hexutil.Big     `json:"value"`
 	Data                 hexutil.Bytes   `json:"data"`
 	Input                hexutil.Bytes   `json:"input"`
+	FeePayer             common.Address  `json:"feePayer"`
 
 	// Introduced by AccessListTxType transaction.
 	AccessList *types.AccessList `json:"accessList,omitempty"`
 	ChainID    *hexutil.Big      `json:"chainId,omitempty"`
 }
 
-func (args *CallArgs) GetMsgIngredients() (*common.Address, types.AccessList, []byte) {
-	if args.AccessList != nil {
-		return args.To, *args.AccessList, args.InputData()
-	} else {
-		return args.To, nil, args.InputData()
-	}
-}
-
 func (args *CallArgs) InputData() []byte {
 	if args.Input != nil {
 		return args.Input
@@ -425,22 +413,6 @@ func (s *PublicBlockChainAPI) EstimateGas(ctx context.Context, args CallArgs, bl
 	return DoEstimateGas(ctx, s.b, args, bNrOrHash, s.b.RPCEVMTimeout(), new(big.Int).SetUint64(gasCap))
 }
 
-func deriveCallArgToMsg(b Backend, args CallMsg, header *types.Header, gasCap uint64) (*types.Transaction, error) {
-	var (
-		baseFee       = new(big.Int).SetUint64(params.ZeroBaseFee)
-		to, acl, data = args.GetMsgIngredients()
-	)
-
-	if header.BaseFee != nil {
-		baseFee = header.BaseFee
-	}
-	intrinsicGas, err := types.IntrinsicGas(data, acl, to == nil, b.ChainConfig().Rules(header.Number))
-	if err != nil {
-		return nil, err
-	}
-	return args.ToMessage(gasCap, baseFee, intrinsicGas)
-}
-
 func DoEstimateGas(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.BlockNumberOrHash, timeout time.Duration, gasCap *big.Int) (hexutil.Uint64, error) {
 	var feeCap *big.Int
 	if args.GasPrice != nil {
@@ -449,20 +421,16 @@ func DoEstimateGas(ctx context.Context, b Backend, args CallArgs, blockNrOrHash
 		feeCap = common.Big0
 	}
 
-	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	state, _, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
 	if err != nil {
 		return 0, err
 	}
+
 	balance := state.GetBalance(args.From) // from can't be nil
-	msg, err := deriveCallArgToMsg(b, &args, header, gasCap.Uint64())
-	if err != nil {
-		return 0, err
+	if !common.EmptyAddress(args.FeePayer) {
+		// overide the balance if the `feePayer` is given
+		balance = state.GetBalance(args.FeePayer)
 	}
-	requiredMinBalance := new(big.Int).Mul(new(big.Int).SetUint64(uint64(args.Gas)), msg.EffectiveGasPrice(header, b.ChainConfig()))
-	if balance.Cmp(requiredMinBalance) < 0 {
-		balance = new(big.Int).Add(balance, requiredMinBalance)
-	}
-
 	// Create a helper to check if a gas allowance results in an executable transaction
 	executable := func(gas uint64) (bool, *blockchain.ExecutionResult, error) {
 		args.Gas = hexutil.Uint64(gas)
diff --git a/api/tx_args.go b/api/tx_args.go
index bb6a69de..4276f5d3 100644
--- a/api/tx_args.go
+++ b/api/tx_args.go
@@ -547,6 +547,7 @@ type EthTransactionArgs struct {
 	MaxPriorityFeePerGas *hexutil.Big    `json:"maxPriorityFeePerGas"`
 	Value                *hexutil.Big    `json:"value"`
 	Nonce                *hexutil.Uint64 `json:"nonce"`
+	FeePayer             common.Address  `json:"feePayer"`
 
 	// We accept "data" and "input" for backwards-compatibility reasons.
 	// "input" is the newer name and should be preferred by clients.
@@ -559,14 +560,6 @@ type EthTransactionArgs struct {
 	ChainID    *hexutil.Big      `json:"chainId,omitempty"`
 }
 
-func (args *EthTransactionArgs) GetMsgIngredients() (*common.Address, types.AccessList, []byte) {
-	if args.AccessList != nil {
-		return args.To, *args.AccessList, args.data()
-	} else {
-		return args.To, nil, args.data()
-	}
-}
-
 // from retrieves the transaction sender address.
 func (args *EthTransactionArgs) from() common.Address {
 	if args.From == nil {
```


The fix more simpler and now the API can take optional param, `FeePayer`.
```sh
#!/bin/bash

curl -X 'POST' \
  'http://localhost:8551' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "method": "kaia_estimateGas",
  "id": 1,
  "jsonrpc": "2.0",
  "params": [
    {
      "feePayer": "0x96Bd8E216c0D894C0486341288Bf486d5686C5b6",
      "from": "0x9a8d1dfdad4806605e3c63067897e4f13b06a5b4",
      "to": "0x10235AC9fCb213e1B229f636913637CD35Bf4031",
      "gas": "0x1cbf1",
      "value": "0x0",
"gasPrice":"0x5d21dba00",
      "input": "0xe3dc878300000000000000000000000000000000000000000000000000000000000004d2"
    },
    "0x7bd"
  ]
}'
```

Reply:
But the original example must have worked without feePayer. None of the SDKs will add feePayer.

```
 curl -X 'POST' \
   'http://localhost:8551/' \
   -H 'accept: application/json' \
   -H 'Content-Type: application/json' \
   -d '{
   "method": "kaia_estimateGas",
   "id": 1,
   "jsonrpc": "2.0",
   "params": [
     {
       "from": "0x9a8d1dfdad4806605e3c63067897e4f13b06a5b4",
       "to": "0x10235AC9fCb213e1B229f636913637CD35Bf4031",
       "gas": "0x1cbf1",
       "value": "0x0",
       "gasPrice":"0x5d21dba00", # `insufficient fund..` error occurs if this field was present
       "input": "0xe3dc878300000000000000000000000000000000000000000000000000000000000004d2"
     }
   ]
 }'
 ```

Reply:
Let's clear the problem and the desired solution so that I can catch your thought more clearly and make a consensus.

Problem. `insufcicent fund...` error occurs if no sender has enough balance to simulate the transaction.
Solution.
1. (PR's solution-1): Add required minimum balance to the `sender` initially
- implication: No SDK change required
- note: Ambiguous. Because it checks the balance of the `sender`, not `feePayer`. Seems to just be making working one anyhow
2. (PR's solution-2): Add a `FeePayer` property newly.
- implication: SDK should be aware of this new property, BUT no existing behavior changed by this new property.


I just started from solution 1, but I am now thinking of transit to solution 2. I described the approach, implication, and note above. PTAL. Please give your opinion and also put an alternative if there's no desired solution is not listed above. Thanks

Reply:
Sorry, we cannot make SDK to enter `feePayer` field to call `kaia_estimateGas`. EstimateGas does not need to be a precise simulation of the transaction. Calculating the contract execution part should be enough. This API is used to enter the gasLimit in a new Tx, not predicting the outcome of the transaction. "just be making working anyhow" is what I want.

Reply:
Okay. Let's talk at TMR scrum and complete the implementation. Thanks